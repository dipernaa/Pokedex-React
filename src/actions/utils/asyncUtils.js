export function buildActionNames(baseName) {
  return {
    failure: `${baseName}_FAILURE`,
    request: `${baseName}_REQUEST`,
    success: `${baseName}_SUCCESS`,
  };
}

export function buildActions(actionNames) {
  return {
    failure: (err) => ({
      type: actionNames.failure,
      ...err
    }),
    request: () => ({
      type: actionNames.request
    }),
    success: (data) => ({
      type: actionNames.success,
      ...data
    }),
  };
}

export function createAsyncAction(actionNames, asyncFn) {
  const actions = buildActions(actionNames);

  const actionCreator = (dispatch, getState) => {
    dispatch(actions.request());
    const state = getState();

    return asyncFn(state)
      .then(result => {
        dispatch(actions.success(result));
        return {
          ok: true,
          data: result
        };
      })
      .catch(err => {
        dispatch(actions.failure(err));
        return {
          ok: false,
          data: err
        };
      });
  };

  return actionCreator;
}

export default {
  buildActionNames,
  createAsyncAction
};
