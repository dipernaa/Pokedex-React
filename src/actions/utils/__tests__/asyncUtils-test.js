import mockStore from 'redux-mock-store';
import { buildActionNames, buildActions, createAsyncAction } from '../asyncUtils';

const mockReduxStore = mockStore();

beforeEach(() => {
  mockReduxStore.clearActions();
});

it('builds action names from a base name', () => {
  expect(buildActionNames('TEST_BASE'))
    .toMatchSnapshot();
});

it('builds success, request, and failure functions', () => {
  const actionNames = buildActionNames('TEST_BASE');
  const actions = buildActions(actionNames);

  expect(actions).toMatchSnapshot();
  expect(actions.request()).toMatchSnapshot();
  expect(actions.success({ value: 'data 1' })).toMatchSnapshot();
  expect(actions.failure({ error: 'error 1' })).toMatchSnapshot();
});

it('dispatches the correct actions for a failed async action', () => {
  const actionNames = buildActionNames('TEST_BASE');
  const asyncFn = () => Promise.reject({ error: 'error 2' });

  return mockReduxStore.dispatch(createAsyncAction(actionNames, asyncFn))
    .then(() => {
      expect(mockReduxStore.getActions()).toMatchSnapshot();
    });
});

it('dispatches the correct actions for a successful async action', () => {
  const actionNames = buildActionNames('TEST_BASE');
  const asyncFn = () => Promise.resolve({ data: 'data 1' });

  return mockReduxStore.dispatch(createAsyncAction(actionNames, asyncFn))
    .then(() => {
      expect(mockReduxStore.getActions()).toMatchSnapshot();
    });
});
