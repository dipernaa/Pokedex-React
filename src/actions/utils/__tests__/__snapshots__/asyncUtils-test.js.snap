// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds action names from a base name 1`] = `
Object {
  "failure": "TEST_BASE_FAILURE",
  "request": "TEST_BASE_REQUEST",
  "success": "TEST_BASE_SUCCESS",
}
`;

exports[`builds success, request, and failure functions 1`] = `
Object {
  "failure": [Function],
  "request": [Function],
  "success": [Function],
}
`;

exports[`builds success, request, and failure functions 2`] = `
Object {
  "type": "TEST_BASE_REQUEST",
}
`;

exports[`builds success, request, and failure functions 3`] = `
Object {
  "type": "TEST_BASE_SUCCESS",
  "value": "data 1",
}
`;

exports[`builds success, request, and failure functions 4`] = `
Object {
  "error": "error 1",
  "type": "TEST_BASE_FAILURE",
}
`;

exports[`dispatches the correct actions for a failed async action 1`] = `
Array [
  Object {
    "type": "TEST_BASE_REQUEST",
  },
  Object {
    "error": "error 2",
    "type": "TEST_BASE_FAILURE",
  },
]
`;

exports[`dispatches the correct actions for a successful async action 1`] = `
Array [
  Object {
    "type": "TEST_BASE_REQUEST",
  },
  Object {
    "data": "data 1",
    "type": "TEST_BASE_SUCCESS",
  },
]
`;
